mlflow/langchain/__init__.py

```diff
@@ -393,14 +575,62 @@ def _save_model(model, path):
     return model_data_kwargs
 
 
-def _load_model(path, agent_path=None, tools_path=None, agent_primitive_path=None):
-    model = None
-    if agent_path is None and tools_path is None:
-        from langchain.chains.loading import load_chain
+def _load_from_pickle(loader_fn_path, persist_dir):
+    if loader_fn_path is None:
+        raise mlflow.MlflowException.invalid_parameter_value(
+            "Missing file for loader_fn which is required to build the model."
+        )
+    with open(loader_fn_path, "rb") as f:
+        loader_fn = cloudpickle.load(f)
+    return loader_fn(persist_dir)
+
+
+def _load_model(
+    path,
+    model_type,
+    agent_path=None,
+    tools_path=None,
+    agent_primitive_path=None,
+    loader_fn_path=None,
+    persist_dir=None,
+):
+    from langchain.chains.loading import load_chain
+    from langchain.chains import (
+        RetrievalQA,
+        VectorDBQA,
+        VectorDBQAWithSourcesChain,
+        APIChain,
+        HypotheticalDocumentEmbedder,
+        SQLDatabaseChain,
+    )
+
+    unserializable_object_name_map = {
+        RetrievalQA.__name__: "retriever",
+        VectorDBQA.__name__: "vectorstore",
+        VectorDBQAWithSourcesChain.__name__: "vectorstore",
+        APIChain.__name__: "requests_wrapper",
+        HypotheticalDocumentEmbedder.__name__: "embeddings",
+        SQLDatabaseChain.__name__: "database",
+    }
 
+    model = None
+    if model_type in (
+        RetrievalQA.__name__,
+        VectorDBQA.__name__,
+        VectorDBQAWithSourcesChain.__name__,
+        APIChain.__name__,
+        HypotheticalDocumentEmbedder.__name__,
+        SQLDatabaseChain.__name__,
+    ):
+        kwargs = {
+            unserializable_object_name_map[model_type]: _load_from_pickle(
+                loader_fn_path, persist_dir
+            )
+        }
```

Comment:
/ai simplify this code using `unserializable_object_name_map.get(model_type)` and assignment expression.